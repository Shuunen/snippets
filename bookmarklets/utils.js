/* eslint-disable-next-line no-unused-vars */
class Shuutils {
  constructor (app) {
    app = app || {
      id: 'shu-app',
      debug: true
    }
    this.app = app
    this.accentsIn = 'ÀÁÂÃÄÅĄàáâãäåąßÒÓÔÕÕÖØÓòóôõöøóÈÉÊËĘèéêëęðÇĆçćÐÌÍÎÏìíîïÙÚÛÜùúûüÑŃñńŠŚšśŸÿýŽŻŹžżź'
    this.accentsOut = 'AAAAAAAaaaaaaaBOOOOOOOOoooooooEEEEEeeeeeeCCccDIIIIiiiiUUUUuuuuNNnnSSssYyyZZZzzz'
  }

  log (...stuff) {
    if (!this.app.debug) {
      return
    }
    stuff.unshift(this.app.id + ' :')
    console.log.apply(console, stuff)
  }

  warn (...stuff) {
    stuff.unshift(this.app.id + ' :')
    console.warn.apply(console, stuff)
  }

  error (...stuff) {
    stuff.unshift(this.app.id + ' :')
    console.error.apply(console, stuff)
  }

  readableString (str) {
    return str.split('').map(function (letter) {
      var i = this.accentsIn.indexOf(letter)
      return i !== -1 ? this.accentsOut[i] : letter
    }).join('').replace(/(\W|-)/gi, ' ').replace(/\s+/, ' ')
  }

  debounce (func, wait, immediate) {
    var timeout
    return function () {
      var context = this
      var args = arguments
      var later = function later () {
        timeout = null
        if (!immediate) {
          func.apply(context, args)
        }
      }
      var callNow = immediate && !timeout
      clearTimeout(timeout)
      timeout = setTimeout(later, wait)
      if (callNow) {
        func.apply(context, args)
      }
    }
  }

  findOne (selector, context, dontYell) {
    context = context || document
    var item = context.querySelector(selector)
    if (item && this.app.debug) {
      this.log('found element matching "' + selector + '"')
    } else if (!item && !dontYell) {
      this.warn('found no element for selector "' + selector + '"')
    }
    return item
  }

  findFirst (selector, context, dontYell) {
    return this.findAll(selector, context, dontYell)[0]
  }

  findAll (selector, context, dontYell) {
    if (!selector || !selector.length || selector.length === 1) {
      this.error('incorrect selector : ', selector)
    }
    context = context || document
    var items = Array.prototype.slice.call(context.querySelectorAll(selector))
    if (items.length && this.app.debug) {
      this.log('found', items.length, 'elements matching "' + selector + '"')
    } else if (!items.length && !dontYell) {
      this.warn('found no elements for selector "' + selector + '"')
    }
    return items
  }
}
