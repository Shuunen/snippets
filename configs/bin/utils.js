import { copyFile, mkdir } from 'node:fs/promises'
import path from 'node:path'

/**
 * Get the filename from a filepath
 * @param {string} filepath the filepath to get the filename from
 * @returns {string} the filename
 * @example filename('C:\\Users\\me\\file.txt') // 'file.txt'
 */
export function filename(filepath = '') {
  return /[/\\](?<name>[\w.-]+)$/u.exec(filepath)?.groups?.name ?? ''
}

/**
 * Remove lines matching a regexp list
 * @param {string} content the content to clean
 * @param {RegExp[]} regexList the regex list to match
 * @returns {string} the cleaned content
 */
export function removeLinesMatching(content, regexList) {
  const lines = content.split('\n')
  const filteredLines = lines.filter(line => !regexList.some(regex => regex.test(line)))
  return filteredLines.join('\n').trim()
}

/**
 * Remove lines after a given regex
 * @param {string} content the content to remove lines from
 * @param {RegExp} regex the regexp to match the line after which to remove content
 * @returns {string} the content without the lines after the matching line
 */
export function removeLinesAfter(content, regex) {
  const lines = content.split('\n')
  const index = lines.findIndex(line => regex.test(line))
  if (index === -1) return content.trim()
  return lines.slice(0, index).join('\n').trim()
}

/**
 * Convert carriage return to unix line endings
 * @param {string} content the content to be processed
 * @returns the processed content with unix line endings
 */
export function useUnixCarriageReturn(content) {
  return content.replace(/\r\n/gu, '\n')
}

/**
 * Clean a file details content
 * @param {string} content the file content to clean
 * @param {RegExp} [linesAfter] a regex to remove lines after
 * @param {RegExp[]} [linesMatching] a list of regex to remove lines matching
 * @param shouldClearSpaces if true will also clear spaces
 * @returns {string} the cleaned file content
 */
// eslint-disable-next-line @typescript-eslint/max-params
export function clean(content, linesAfter, linesMatching, shouldClearSpaces = true) {
  if (!content) return ''
  let output = content
  if (linesAfter) output = removeLinesAfter(output, linesAfter)
  if (linesMatching) output = removeLinesMatching(output, linesMatching)
  if (shouldClearSpaces) output = output.replace(/\s*/gu, '')
  return output
}

/* c8 ignore start */

/**
 * Normalize a filepath with slash style
 * @param {string} filepath the filepath
 * @param {boolean} shouldUseTilde use tilde will replace the home directory with ~
 * @param {string} home the home directory path
 * @returns the normalized path
 */
export function normalizePathWithSlash(filepath, shouldUseTilde = false, home = process.env.HOME ?? '') {
  let outPath = path.normalize(filepath).replace(/\\/gu, '/')
  if (shouldUseTilde) outPath = outPath.replace(normalizePathWithSlash(home), '~')
  return outPath
}

/**
 * Copy a file
 * @param {string} source the source file
 * @param {string} destination the destination file
 * @returns {Promise<boolean>} some bool result; i dont know im in the train to Paris
 */
export async function copy(source, destination) {
  // destination will be created or overwritten by default.
  const destinationFolder = destination.replace(filename(destination), '')
  // eslint-disable-next-line @typescript-eslint/naming-convention
  await mkdir(destinationFolder, { recursive: true })

  return copyFile(source, destination)
    .then(() => true)
    .catch(error => {
      console.log(error)
      return false
    })
}
/* c8 ignore stop */
