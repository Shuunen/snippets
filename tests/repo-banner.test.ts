import { logger, replaceAndCheck, replaceAndCheckById } from '../one-file/repo-banner.utils'
import { svgMockA } from './repo-banner.mock'
import { check, checkSnapshot } from './utils'

check('replaceAndCheck A', replaceAndCheck('Hello world !', /(?<before>Hello )world(?<after> !)/gu, 'there'), 'Hello there !')
logger.info('Dont worry, the following error is expected')
check('replaceAndCheck B generate error', replaceAndCheck('Hello world !', /(?<before>Hello )world(?<after> !)/gu, 'world'), 'Hello world !')
logger.info('That was it, are you good ?')

checkSnapshot('replaceAndCheckById A', replaceAndCheckById(svgMockA, 'projectName', 'My project name'))
