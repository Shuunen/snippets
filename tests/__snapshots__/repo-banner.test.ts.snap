// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`replaceAndCheckById A 1`] = `
"<svg viewBox=\\"0 0 388 240\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\">
<defs>
  <linearGradient gradientUnits=\\"userSpaceOnUse\\" x1=\\"200\\" y1=\\"150\\" x2=\\"200\\" y2=\\"200\\" id=\\"triangle-gradient\\" >
    <stop offset=\\"0\\" style=\\"stop-color: #ffffff\\"/>
    <stop offset=\\"1\\" style=\\"stop-color: #ffffff30\\"/>
  </linearGradient>
</defs>
<rect width=\\"388\\" height=\\"240\\" style=\\"fill: rgb(14, 17, 23);\\">
  <title>github-background</title>
</rect>
<g  transform=\\"matrix(0.9, 0, 0, 0.9, 19, 14)\\">
  <title>header</title>
  <path d=\\"M 99.658 101.782 C 99.658 96.029 Z\\" style=\\"fill: rgb(229, 229, 229);\\">
    <title>github-logo</title>
  </path>
  <text style=\\"fill: rgb(229, 229, 229); font-family: Arial, sans-serif; font-size: 33px\\" x=\\"182\\" y=\\"92\\">
    <title>plus</title>+
  </text>
  <path d=\\"M 236.122 30 L 308 Z\\" style=\\"fill: rgb(229, 229, 229);\\">
    <title>typescript-logo</title>
  </path>
</g>
<foreignObject x=\\"0\\" y=\\"150\\" width=\\"100%\\" height=\\"30\\" style=\\"font-size: 28px; color: white; text-align: center; font-family: Arial;\\">
  <xhtml:div id=\\"projectScope\\" style=\\"display: inline-block;\\">Shuunen</xhtml:div>
  <xhtml:div> / </xhtml:div>
  <xhtml:div id=\\"projectName\\" style=\\"display: inline-block; font-weight: 700;\\">My project name</xhtml:div>
</foreignObject>
<foreignObject id=\\"projectDescription\\" x=\\"0\\" y=\\"190\\" width=\\"100%\\" height=\\"20\\" style=\\"font-size: 16px; color: white; text-align: center; font-family: Arial; letter-spacing: 0.5px;\\">The description is kind of pretty !</foreignObject>
<path d=\\"M 388.377 -3.102 173.477\\" fill=\\"none\\" stroke-width=\\"1.5\\" style=\\"stroke: rgb(255, 255, 255); visibility: hidden;\\">
  <title>triangle-guides</title>
</path>
</svg>"
`;
